### PR TITLE
Integrate bitcoin_hashes, remove rust-crypto and fuzz_util

### PR DESCRIPTION
**Pull request by dongcarl**

### CODE REVIEW

**Comment by dongcarl** [comment]
Attention Reviewers: I have identified the cause of the panic in Travis and fixed it in https://github.com/rust-bitcoin/bitcoin_hashes/pull/29.

File: src/blockdata/block.rs [hunk 1]
```diff
@@ -120,11 +122,15 @@ impl BlockHeader {
     /// is correct, but does not verify that the transactions are valid or encoded
     /// correctly.
     pub fn spv_validate(&self, required_target: &Uint256) -> Result<(), util::Error> {
+        use std::mem;
```
[the remark this answers is not in the document]
Comment (sgeisler) [reply src/blockdata/block.rs]: I'm doing this too if the usage of the symbol is mostly local. It also makes moving functions between modules much easier.

Do you see any problems with doing it that way or is it just unusual?

File: src/blockdata/block.rs [hunk 2]
```diff
@@ -120,11 +122,15 @@ impl BlockHeader {
...
         let target = &self.target();
         if target != required_target {
             return Err(SpvBadTarget);
         }
-        let hash = &self.bitcoin_hash().into_le();
+        let data: [u8; 32] = self.bitcoin_hash().into_inner();
+        let mut ret: [u64; 4] = unsafe { mem::transmute(data) };
```
[the remark this answers is not in the document]
Comment (dongcarl) [reply src/blockdata/block.rs]: Unfortunately not for our versions. Also, `from_ne_bytes` uses `unsafe` the exact same way we do: https://doc.rust-lang.org/src/core/num/mod.rs.html#3853-3855

[the remark this answers is not in the document]
Comment (sgeisler) [reply src/blockdata/block.rs]: I second that, unless there is a benchmark showing a substantial speedup we should always avoid `unsafe`. The transmute here is probably also platform dependent (le/be integer encoding).

File: src/blockdata/block.rs [hunk 3]
```diff
@@ -120,11 +122,15 @@ impl BlockHeader {
     /// is correct, but does not verify that the transactions are valid or encoded
     /// correctly.
     pub fn spv_validate(&self, required_target: &Uint256) -> Result<(), util::Error> {
+        use std::mem;
         let target = &self.target();
         if target != required_target {
```
[the remark this answers is not in the document]
Comment (dongcarl) [reply src/blockdata/block.rs]: It doesn't make a huge difference here, as dereferencing `required_target` will call `Copy`, and that is arguably even more expensive.

**Comment by dongcarl** [comment]
Going to push rebased version with `bitcoin_hashes` bumped to `0.3.0`.

File: src/blockdata/block.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgeisler) [reply src/blockdata/block.rs]: Why would you use `from_ne_bytes` ("native endianess") in our case?

[the remark this answers is not in the document]
Comment (dongcarl) [reply src/blockdata/block.rs]: Ah. Good catch. Lemme fix.

File: src/blockdata/block.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/blockdata/block.rs]: We just don't do that for rust-lightning and never(rarely?) see it in general rust projects. But yeah function moving between modules is a good point, maybe you should stick to symbol strictly local ?

File: src/blockdata/block.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ariard) [reply src/blockdata/block.rs]: Yes was a nit, just to be fixed, that something happening elsewhere in rust-bitcoin

**Comment by dongcarl** [comment]
Fixed some minor feature naming so that Travis is happy.

File: src/internal_macros.rs [hunk 4]
```diff
@@ -299,7 +280,7 @@ macro_rules! display_from_debug {
 macro_rules! hex_script (($s:expr) => (::blockdata::script::Script::from(::hex::decode($s).unwrap())));
 
 #[cfg(test)]
-macro_rules! hex_hash (($s:expr) => (::util::hash::Sha256dHash::from(&::hex::decode($s).unwrap()[..])));
+macro_rules! hex_hash (($s:expr) => (::bitcoin_hashes::Hash::from_slice(&::hex::decode($s).unwrap()).unwrap()));
```
Comment (TheBlueMatt) [inline src/internal_macros.rs]: You can replace the bitcoin_hashes::Hash with bitcoin_hashes::sha256d::Hash::from_slice here to make it match the old version.

**Comment by tamasblummer** [comment]
This is rather disruptive downstream. Maybe we should add an implementation of Sha256dHash that wraps sha256d::Hash and implements the most used methods, that are in my projects: Sha256dHash::default ::from ::from_data ::from_hex ::as_bytes

**Comment by TheBlueMatt** [comment]
I think we generally do a major version bump almost-by-itself for a dep bump like this so that people can update at their leisure. If you need it downstream we could maybe land #218 and a few others in their own version first, but I'd rather not add wrappers to be removed in the next version :/.

**Comment by dongcarl** [comment]
@sgeisler Please re-ACK, thanks!

**Comment by dongcarl** [comment]
@tamasblummer I agree with @TheBlueMatt, I'm happy to rebase over your PRs if we wanna do a version with those first! :smile:

**Pull request merged**